### ISSUE TITLE
[3DS] Libretro Mame 2003-Plus: Can't save controller configs per-game (Save game remap/Save core remap don't seem to work)

### ISSUE THREAD

**Issue by sunrisepi** [libretro/RetroArch#10886]
## Description
Within Quick Menu->Controls->Port 1 Controls, saving controller changes via Save Game Remap File and Save Core Remap File don't do anything. Only global settings (for the entire core) via Settings->Input->Port 1 Controls work. However, since MAME games use different layouts (some are vertical orientation (example: Pac-Man), others horizontal (example: Joust)), this doesn't work well. The way this works now, the user has to reconfigure their global settings every time a game requires a different orientation.

Based on https://docs.libretro.com/guides/arcade-getting-started/ and https://docs.libretro.com/library/mame2003_plus/, I think this should be working. I chose the core MAME 2003-Plus as it seemed it would support remapping controller input.

I also had this issue occur on Mame2003 core. I had opened the issue first against mame-2003-plus but as it was asked if this happened on another core, and it did, perhaps this is a RetroArch issue:  https://github.com/libretro/mame2003-plus-libretro/issues/819

Would appreciate knowing how to fix this... tried searching for hours on gbatemp, github, general internet and didn't find any suggestions to fix this specific issue for 3DS RetroArch.

### Expected behavior

Ability to save controller settings per game. That way when I play Pac-Man I can use one controller config, and in Joust, I could use another.

### Actual behavior

Try setting control configs via Save game remap and they don't work.

### Steps to reproduce the bug

I've tried following https://docs.libretro.com/guides/input-and-controls/#remapping-controls-for-individual-cores-or-content.

Navigated via my defined hotkey to Quick Menu -> Controls -> Port 1 Controls (this seems to be the only available option for 3DS).

This specifies by default "D-Pad Up -1 (N/A) = Down," "D-Pad Down +1 (N/A) = Down," "D-Pad Left -0 (N/A) = Left" and "D-Pad Right +0 (N/A) = Right".

I also see "Left Analog X+/X-/Y+/Y-" (all have "(N/A)" in the name as well) set to nothing ("--") and "Auto: C Stick Right/Left/Down/Up" set to nothing ("--").

I have tried modifying the settings for D-Pad controls, Left Analog controls and C Stick controls. Nothing changes.

I have tried setting everything to "--" for D-Pad, Left Analog & C Stick. Nothing changes. The default controls for the circle pad are still in place and not changed, so that when the 3DS is rotated counter-clockwise, pushing up on the circle pad = going right (instead of when I want pac-man to move up, based on current screen orientation, with screen on the left in vertical orientation). Other controls are also similarly unchanged.

Tried both "Save Core Remap File" and Save Game Remap File," also reloading both, no change, stays the same.

The only thing that seems to do anything is when I set "Left Analog" controls to anything (such as Left, Right, Down, Up), it breaks the circle pad control, and no direction works. You would think to fix it, you'd change it back to what it was before - setting "Left Analog" to "--" for every direction (L/R/D/U) & saving remap file. But that doesn't work and the circle pad remains broken. To fix it I had to select Quick Menu -> Controls -> Delete Core Remap File. Then the circle pad reverts to the same undesired configuration, where the control pad is set as if the 3DS is being held in the horizontal orientation.

I then discovered it seems you can only update this globally for the entire Mame-2003-plus core, not by-game within the core... Global settings work but this isn't helpful when maybe 80% of games have vertical orientation and 30% have horizontal. Due to this you need to manually update the global config when you switch between games.

### Bisect Results

I just started using MAME 2003-plus on 3DS yesterday so don't know how long this may have been broken.

### Version/Commit
Version 1.8.9, from https://buildbot.libretro.com/stable/1.8.9/nintendo/3ds/RetroArch_cia.7z

- RetroArch: [version/commit]

### Environment information

Using Nintendo CDS with CFW

Thanks!

**Comment by sunrisepi** [comment]
@WindyWinston provided the fix for me on the issue I had opened initially in the other section (libretro/mame2003-plus, #819):  Just use the MAME configuration menu!  I really did not see this.  I had to dig for this feature, buried in the menu options.  But when I found it, the fix was clear.

How to configure: Once a core & game are loaded, use the hotkey to bring up the RetroArch menu, then select **Quick Menu->Options->Display MAME menu** and set that to **ON.** Then select **Input (this game)** and for each action used by the game (such as P1 Left, P1 Right, P1 Button 1, etc.) you select it & then press the button on the controller you'd like to map to the action.

That works well & tried this for 2 games: Joust & Berzerk. I also verified upon reloading each of these games that the settings I had configured before were retained.

So, to summarize:  Within the MAME 2003-plus core, I set up global options for the controller **(Settings->Input->Port 1 Controls**) for vertical orientation games, as the majority of the games I play in MAME are of that configuration.  For any that I discover which are horizontal orientation, I will set up per-game controller configurations for them using the **MAME menu.**

I think it would be really helpful if we could have documentation for 3DS users of RetroArch & MAME cores, so they don't fall into the same trap I did.  Is this documentation already in progress?  If not, I'd be happy to help create some.  The documentation I found at https://docs.libretro.com/guides/install-3ds2ds/ was very helpful in getting RetroArch up and running on my 3DS; just a few more paragraphs on how to improve RetroArch load times (only have the necessary cores in the cores directory to speed up load time, for example) and configure controllers in MAME could save folks quite a bit of time.

Great to finally have MAME set up and working well now!  RetroArch & MAME is a lot of fun.  Thanks! 😉 🥳
